clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/handler/KubernetesJobHandler.java

```diff
@@ -0,0 +1,95 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.clouddriver.kubernetes.v2.op.handler;
+
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.artifact.ArtifactReplacerFactory;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.caching.agent.KubernetesCacheDataConverter;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.caching.agent.KubernetesJobCachingAgent;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.caching.agent.KubernetesV2CachingAgent;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.description.KubernetesSpinnakerKindMap;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.description.manifest.KubernetesKind;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.description.manifest.KubernetesManifest;
+import com.netflix.spinnaker.clouddriver.model.Manifest.Status;
+import io.kubernetes.client.models.V1Job;
+import io.kubernetes.client.models.V1JobSpec;
+import io.kubernetes.client.models.V1JobStatus;
+import org.springframework.stereotype.Component;
+
+@Component
+public class KubernetesJobHandler extends KubernetesHandler implements
+  CanDelete {
+
+  public KubernetesJobHandler() {
+    registerReplacer(ArtifactReplacerFactory.dockerImageReplacer());
+    registerReplacer(ArtifactReplacerFactory.configMapVolumeReplacer());
+    registerReplacer(ArtifactReplacerFactory.secretVolumeReplacer());
+    registerReplacer(ArtifactReplacerFactory.configMapEnvFromReplacer());
+    registerReplacer(ArtifactReplacerFactory.secretEnvFromReplacer());
+  }
+
+  @Override
+  public KubernetesKind kind() {
+    return KubernetesKind.JOB;
+  }
+
+  @Override
+  public boolean versioned() {
+    return false;
+  }
+
+  @Override
+  public KubernetesSpinnakerKindMap.SpinnakerKind spinnakerKind() {
+    return KubernetesSpinnakerKindMap.SpinnakerKind.SERVER_GROUPS;
+  }
+
+  @Override
+  public Status status(KubernetesManifest manifest) {
+    V1Job v1Job = KubernetesCacheDataConverter.getResource(manifest, V1Job.class);
+    return status(v1Job);
+  }
+
+  @Override
+  public Class<? extends KubernetesV2CachingAgent> cachingAgentClass() {
+    return KubernetesJobCachingAgent.class;
+  }
+
+  private Status status(V1Job job) {
+    Status result = new Status();
+    V1JobStatus status = job.getStatus();
+    if (status == null) {
+      result.unstable("No status reported yet")
+          .unavailable("No availability reported");
+      return result;
+    }
+
+    int completions = 1;
+    V1JobSpec spec = job.getSpec();
+    if (spec != null && spec.getCompletions() != null) {
+      completions = spec.getCompletions();
+    }
+    int succeeded = 0;
+    if (status.getSucceeded() != null) {
+      succeeded = status.getSucceeded();
+    }
+    if (succeeded < completions) {
+      return result.unstable("Waiting for jobs to finish");
```

Comment:
Seems the only way this stage can fail is when the orca/deck configured time out is hit? It would be nice to somehow tie this into [active deadline seconds](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#job-termination-and-cleanup), or introduce a new concept of stability that could cause a stage to fail immediately. The latter could help when deploying individual pods that exit unsuccessfully.

Reply:
Yes, that's correct.

I had a version that tried to make use of active deadline. It would be nice, yes. Seems like orca is in charge of failing stages and there's no good way for clouddriver to signal a failure? My attempt at throwing an exception to abort did not get propagated to orca.

That's why I went looking for the timeout option instead, which kind of works but it's clumsy and sub-optimal.

```java
    if (succeeded < completions) {
      Long startTimeMillis = status.getStartTime().getMillis();
      Long deadlineSeconds = spec.getActiveDeadlineSeconds();
      if (startTimeMillis != null && deadlineSeconds != null &&
        startTimeMillis + Duration.ofSeconds(deadlineSeconds).toMillis() < System.currentTimeMillis()) {
        return result.unstable(
          "Job " + job.getMetadata().getName() + " in namespace " +
          job.getMetadata().getNamespace() + " exceeded deadline");
      }
      return result.unstable("Waiting for all jobs to finish");
    }
```

Reply:
We could also try something like

`result.failed("Job timed out")` which orca treats as an immediate failure (regardless of timeout/retries).

Reply:
Sounds like a plan, would you like me to do that as part of this PR or in another one after this is merged?

Reply:
Let's do it in a separate PR - it will require orca support too.